File: typescript/src/pipelines/zebraWings/embeddings/CombineNgramsLayer.ts
import * as tf from '@tensorflow/tfjs';
import { l2Normalize } from '@tensorflow/tfjs-layers/dist/losses';

// given an imput composed of max_ngrams x 300d this layer will sum
// and normalize all the max_ngrams to get a unique 300d vector representation
export class CombineNgramsLayer extends tf.layers.Layer {
    public static className = 'CombineNgramsLayer';
    public className = CombineNgramsLayer.className;
    // The output shape removes the ngram dimension
    public computeOutputShape(inputShape: number[]) {
        return [inputShape[0], inputShape[1], inputShape[inputShape.length - 1]];
    }
    public call(inputs: tf.Tensor[], kwargs: any) {
        return tf.tidy(() => {
            this.invokeCallHook(inputs, kwargs);
            const combined = inputs[0].sum(2);
            const output = l2Normalize(combined, 2);
            combined.dispose();
            return output;
        });
    }
}

tf.serialization.SerializationMap.register(CombineNgramsLayer);

File: python/src/pipelines/zebra_wings/embeddings/combine_ngrams_layer.py
import keras

# given an imput composed of max_ngrams x 300d this layer will sum
# and normalize all the max_ngrams to get a unique 300d vector representation
class CombineNgramsLayer(keras.layers.Layer):
    def __init__(self, **kwargs):
        if 'input_shape' not in kwargs and 'input_dim' in kwargs:
            kwargs['input_shape'] = (kwargs.pop('input_dim'),)
        super(CombineNgramsLayer, self).__init__(**kwargs)
        self.supports_masking = True

    def call(self, ngram_sequence_identifiers):
        combined = keras.backend.sum(
            ngram_sequence_identifiers, axis=2
        )
        return keras.backend.l2_normalize(combined, axis=2)

    def compute_output_shape(self, input_shape):
        return (None, input_shape[1], input_shape[-1])
